python/graphstorm/distiller/distiller.py

```diff
@@ -0,0 +1,351 @@
+            Data provider for validation data.
+        global_step : int
+            Global step of the model checkpoint.
+        """
+        model.eval()
+        index = 0
+        batch_index = 0
+        total_loss = 0
+        while True:
+            dataset_iterator = eval_dataset_provider.get_iterator()
+            if dataset_iterator is None:
+                break
+            print (f"Eval {index + 1}-th shard by trainer {self.rank}")
+            with th.no_grad():
+                for batch_num, batch in enumerate(dataset_iterator):
+                    batch = self.to_device(batch, self.device)  # Move to device
+                    loss = model.module(batch["input_ids"],
+                        batch["attention_mask"],
+                        batch["labels"])
+                    total_loss += loss.item()
+                    batch_index += 1
+            index += 1
+
+        mean_total_loss = total_loss / batch_index
+        print (f"Eval MSE at step {global_step}: {mean_total_loss}")
+        model.train()
+        eval_dataset_provider.release_iterator()
+
+    def train_shard(
+        self, 
+        global_step, 
+        model, 
+        optimizer, 
+        train_dataset_provider, 
+        eval_dataset_provider, 
+        saved_path,
+        save_model_frequency,
+        eval_frequency,
+        max_global_step,
+    ):
+        """
+        Train using one shard from train_dataset_provider.
+        Parameters
+        ----------
+        global_step : int
+            Global step of the model checkpoint.
+        model : GSDistilledModel
+            Distilled student model.
+        optimizer : torch optimizer
+            optimizer for distillation.
+        train_dataset_provider : DataProvider
+            Data provider for training data.
+        eval_dataset_provider : DataProvider
+            Data provider for validation data.
+        saved_path : str
+            Path to save the model.
+        save_model_frequency : int,
+            Interval to save the model checkpoint.
+        eval_frequency : int,
+            Interval for evaluation.
+        max_global_step : int,
+            Maximum steps for distillation training.
+    
+        Returns
+        -------
+        bool : whether to stop distillation.
+        int : Global step of the model checkpoint.
+        """
+        dataset_iterator = train_dataset_provider.get_iterator()
+        if not dataset_iterator:
+            raise RuntimeError("No training data. Check training dataset as the data sampler"
+                                    "is supposed to be infinite.")
+
+        # TODO (HZ): support prefetch to speed up the training
+        model.train()
+        shard_loss = 0
+        complete = False
+
+        for batch_num, batch in enumerate(dataset_iterator):
+            try:
+                batch = self.to_device(batch, self.device)  # Move to device
+
+                loss = model(batch["input_ids"],
```

Comment:
There is no need to implement a to_device function.

Reply:
The to_device function has some check logic. Here I meant to leave some flexibility for the batch format so that we can add more features in the future easily.

Reply:
This introduces extra overhead of maintaining a universal to_device function, while the to device ops here are quite simple.

Reply:
updated.